.gitlab-ci.yml

```diff
@@ -28,6 +28,451 @@
+#  └── pipelines
+#       ├── lassen.yml
+#       │    ├── <jobs>
+#       │    │    ├── .on_lassen
+#       │    │    ├── .lassen_reproducer_init
+#       │    │    ├── .lassen_reproducer_vars
+#       │    │    ├── .lassen_reproducer_job
+#       │    │    ├── .lassen_job_command
+#       │    │    └── .job_on_lassen
+#       │    ├── <used jobs>
+#       │    │    └── .custom_job  [from .gitlab/custom-jobs-and-variables.yml]
+#       │    └── <used variables>
+#       │         ├── ON_LASSEN
+#       │         ├── ADVANCED_JOB
+#       │         ├── ALL_TARGETS
+#       │         ├── SYS_TYPE
+#       │         ├── LLNL_SERVICE_USER
+#       │         ├── USER
+#       │         ├── GITHUB_PROJECT_NAME
+#       │         ├── GITHUB_PROJECT_ORG
+#       │         ├── LASSEN_JOB_ALLOC
+#       │         └── JOB_CMD
+#       ├── ruby.yml
+#       │    ├── <jobs>
+#       │    │    ├── .on_ruby
+#       │    │    ├── .ruby_reproducer_init
+#       │    │    ├── .ruby_reproducer_vars
+#       │    │    ├── .ruby_reproducer_job
+#       │    │    ├── .ruby_job_command
+#       │    │    ├── .job_on_ruby
+#       │    │    ├── allocate_resources
+#       │    │    └── release_resources
+#       │    ├── <used jobs>
+#       │    │    └── .custom_job  [from .gitlab/custom-jobs-and-variables.yml]
+#       │    ├── <defined variables>
+#       │    │    └── export JOBID
+#       │    └── <used variables>
+#       │         ├── ON_RUBY
+#       │         ├── ADVANCED_JOB
+#       │         ├── ALL_TARGETS
+#       │         ├── SYS_TYPE
+#       │         ├── LLNL_SERVICE_USER
+#       │         ├── USER
+#       │         ├── GITHUB_PROJECT_NAME
+#       │         ├── GITHUB_PROJECT_ORG
+#       │         ├── RUBY_JOB_ALLOC
+#       │         ├── JOB_CMD
+#       │         ├── JOBID
+#       │         ├── ALLOC_NAME
+#       │         └── RUBY_SHARED_ALLOC
+#       └── tioga.yml
+#            ├── <jobs>
+#            │    ├── .on_tioga
+#            │    ├── .tioga_reproducer_init
+#            │    ├── .tioga_reproducer_vars
+#            │    ├── .tioga_reproducer_job
+#            │    ├── .tioga_job_command
+#            │    ├── .job_on_tioga
+#            │    ├── allocate_resources
+#            │    └── release_resources
+#            ├── <used jobs>
+#            │    └── .custom_job  [from .gitlab/custom-jobs-and-variables.yml]
+#            ├── <defined variables>
+#            │    └── PROXY
+#            └── <used variables>
+#                 ├── ON_TIOGA
+#                 ├── ADVANCED_JOB
+#                 ├── ALL_TARGETS
+#                 ├── SYS_TYPE
+#                 ├── LLNL_SERVICE_USER
+#                 ├── USER
+#                 ├── GITHUB_PROJECT_NAME
+#                 ├── GITHUB_PROJECT_ORG
+#                 ├── TIOGA_JOB_ALLOC
+#                 ├── JOB_CMD
+#                 ├── PROXY
+#                 ├── ALLOC_NAME
+#                 └── TIOGA_SHARED_ALLOC
 
```

Comment:
The above can be useful to understand things, and I'm curious to know how it was generated.
However:
- Should it really be version controlled, and inside this file, and at the top of the file... ?
- Maybe it would be interesting to direct people at the actual documentation : https://radiuss-shared-ci.readthedocs.io/en/latest/

Reply:
I manually generated these while trying to understand the workings of the Gitlab CI -- it is distributed among many files and it was not obvious which those files are and what are the connections. 😁 

I personally prefer to keep these notes in this file, otherwise I'll forget where to look for them. These can also be helpful for others trying to understand the MFEM's Gitlab CI inner workings. They are MFEM specific so I'm not sure they will be very useful outside of MFEM. Maybe with some trimming they can be copied to a common documentation file.

Reply:
Can we at least move those lines at the end of the file ? Imagine a cpp file that would start with 500 lines of comments...

I understand the need for it but I am concerned by it to be unmaintainable. Also, it exposes a complexity that is not supposed to be made visible to Shared CI users who should not need to care about how the shared CI manages to be shared, only how to setup the CI.

Feedback on the user guide is welcome.